plugin/include/txn_box/accl_util.h

```diff
@@ -0,0 +1,576 @@
+  using self_type = StringTree<Key, Value>;
+  /// Node layout.
+  struct Node {
+    // types
+    using self_type = Node;
+    using ptr_type  = self_type *;
+
+    Node(Key k, Value v, int32_t r = -1) : key{k}, value{v}, rank{r}
+    {
+      left  = this;
+      right = this;
+    }
+    Key key;
+    Value value;
+    /// bit pos where it differs from previous node.
+    int32_t bit_count{0};
+    /// key/value rank.
+    int32_t rank;
+    /// only two ways, btree
+    self_type *left;
+    self_type *right;
+
+    Node()             = delete;
+    Node(Node const &) = delete;
+    Node(Node &&)      = delete;
+    Node &operator=(Node const &) = delete;
+    Node &operator=(Node &&) = delete;
+  };
+  // types
+  using node_type     = Node;
+  using node_type_ptr = node_type *;
+
+public:
+  // types
+  using value_type = Value;
+  using key_type   = Key;
+
+  /// Construct a new String Tree object
+  StringTree();
+
+  /// We have a bunch of memory to clean up after use.
+  ~StringTree();
+
+  StringTree(StringTree &&)      = delete;
+  StringTree(StringTree const &) = delete;
+  StringTree &operator=(StringTree const &) = delete;
+  StringTree &operator=(StringTree &&) = delete;
+
+  ///
+  /// @brief  Inserts element into the tree, if the container doesn't already contain an element with an equivalent key.
+  /// @return true if the k/v was properly inserted, false if not.
+  ///
+  bool insert(Key const &key, Value const &value, Comparison *cmp = nullptr);
+  ///
+  /// @brief  Finds an element with equivalent key. Only full match.
+  /// @return A pair with a boolean indicating if the key was found and the value associated with it.
+  ///         If value was not found, value can be ignored
+  std::pair<bool, Value> full_match(Key const &key, Comparison *cmp = nullptr) const noexcept;
+  ///
+  /// @brief  Find a value(s) associated with a prefix of a key, this function will perform a prefix search
+  ///         only among all the related keys in the tree.
+  /// @return list of pairs with all the found matches.
+  ///
+  /// TODO: this needs to be changed and use rank to return only 1 value.
+  std::vector<std::pair<Key, Value>> prefix_match(Key const &prefix, Comparison *cmp = nullptr) const;
+
+private:
+  /// hold the first element on the entire tree.
+  node_type_ptr _head;
+  /// this gets incremented on every insert.
+  int32_t _rank_counter;
+  /// recursive memory cleanup function.
+  void freeup(Node *n);
+  /// clean up function, should deal with all the crap.
+  void cleanup();
+};
+
+/// --------------------------------------------------------------------------------------------------------------------
+namespace detail
+{
+// Just to make the code a bit more readable
+static constexpr int BIT_ON{1};
+
+/// @brief return a bit value from a particular position in a byte
+static auto
+get_bit_from_byte(char key, int position) noexcept
+{
+  int const position_in_byte = position - ((position / 8) * 8);
```

Comment:
Why not `position %= 8;`?

Reply:
totally, glad that the compiler does the right thing despite the verbosity of my line.